openhands_aci/tree_sitter/parser.py

```diff
@@ -0,0 +1,138 @@
+import tempfile
+import warnings
+from collections import namedtuple
+from enum import Enum
+from pathlib import Path
+
+from diskcache import Cache
+from grep_ast import filename_to_lang
+from tree_sitter_languages import get_language, get_parser
+
+from openhands_aci.utils.file import get_modified_time, read_text
+from openhands_aci.utils.logger import oh_aci_logger as logger
+
+warnings.filterwarnings('ignore', category=FutureWarning, module='tree_sitter')
+
+ParsedTag = namedtuple(
+    'ParsedTag',
+    ('rel_path', 'abs_path', 'start_line', 'end_line', 'node_content', 'tag_kind'),
+)
+
+
+class TagKind(Enum):
+    DEF = 'def'
+    REF = 'ref'
+    DEF_WITH_BODY = 'def_with_body'
+
+
+class TreeSitterParser:
+    TAGS_CACHE_DIR = '.oh_aci.cache.tags'
+
+    def __init__(self, cache_root_dir: str) -> None:
+        self._load_tags_cache(cache_root_dir)
+
+    def get_tags_from_file(self, abs_path: str, rel_path: str) -> list[ParsedTag]:
+        mtime = get_modified_time(abs_path)
+        cache_key = abs_path
+        cache_val = self.tags_cache.get(cache_key)
+        if cache_val and cache_val.get('mtime') == mtime:
+            return cache_val.get('data')
+
+        data = self._get_tags_raw(abs_path, rel_path)
+        # Update cache
+        self.tags_cache[cache_key] = {'mtime': mtime, 'data': data}
+        return data
+
+    def _get_tags_raw(self, abs_path: str, rel_path: str) -> list[ParsedTag]:
+        lang = filename_to_lang(abs_path)
+        if not lang:
+            return []
+
+        ts_language = get_language(lang)
+        ts_parser = get_parser(lang)
+
+        tags_file_path = (
+            Path(__file__).resolve().parent / 'queries' / f'tree-sitter-{lang}-tags.scm'
+        )
+        if not tags_file_path.exists():
+            return []
+        tags_query = tags_file_path.read_text()
+
+        if not Path(abs_path).exists():
+            return []
+        code = read_text(abs_path)
+        if not code:
+            return []
+
+        parsed_tree = ts_parser.parse(bytes(code, 'utf-8'))
+
+        # Run the tags queries
+        query = ts_language.query(tags_query)
+        captures = query.captures(parsed_tree.root_node)
+
+        parsed_tags = []
```

Comment:
Why do you call it `tags`? Seems a bit confusing? Is it just `node`?

Reply:
It's a term in `tree-sitter` when using query files, the output can be called tag, which is just a named entity, or syntax node: https://tree-sitter.github.io/tree-sitter/code-navigation-systems